libMesh update: Add an optional parameter in executioner for MatSNESMFGetReuseBase.

fdkong opened the pull request:
Some user cases such as contact problems sometime can not reuse the base vector
because the problem configuration may be changed during Jacobian evaluation
so that F(x_n) does not equal to MF(x_n).

Closes #12361

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

fdkong committed:
Add an optional parameter for executioner for MatSNESMFGetReuseBase.

Some user cases such as contact problems sometime can not reuse the base vector
because the problem configuration may be changed during Jacobian evaluation
so that F(x_n) does not equal to MF(x_n).

Closes #12361

fdkong commented:
Waiting for libmesh update

lindsayad commented on framework/src/executioners/Executioner.C:
```diff
@@ -145,6 +149,8 @@ Executioner::Executioner(const InputParameters & parameters)
 
   _fe_problem.getNonlinearSystemBase()._l_abs_step_tol = getParam<Real>("l_abs_step_tol");
+
+  _fe_problem.setSNESMFReuseBase(getParam<bool>("snesmf_reuse_base"));
```
You could add an overload of `FEProblemBase::setSNESMFReuseBase` that takes two arguments, with the second argument specifying whether the parameter was set by the user. So the method you would want to use is [`InputParameters::isParamSetByUser`](http://mooseframework.org/docs/doxygen/moose/InputParameters_8C_source.html#l00793). So here if you go with the overload, you could do:
```c++
_fe_problem.setSNESMFReuseBase(getParam<bool>("snesmf_reuse_base"), parameters.isParamSetByUser("snesmf_reuse_base"));
```

fdkong replied:
I see.

fdkong committed:
Brief summary of changes in this update:

* Fix trailing return type issue affecting GCC 7.1
* Optimization: acquire BC lists outside the CheckPointIO id loop.
* Add --enable-metaphysicl-required configure option.
* Add configure test for std containers with iterating-returning erase.
* Avoid modifying solution vector in TaoOptimizationSolver callbacks.
* Fix warnings discovered in GCC 8.1 and clang 6.1.
* Fix bug in Parallel::minloc/maxloc.
* Add ability to disable default GhostingFunctors.
* Upgrade libMesh MetaPhysicL submodule to version 0.3.1.
* Add new PetscNonlinearSolver interface: set_snesmf_reuse_base().
* Make SparseMatrix::add() const correct.

Refs #000.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/248089/) on 02be9c6 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12362/site)

This comment will be updated on new commits.

jwpeterson commented:
@fdkong, please add the following to your commit message for this change:

Brief summary of changes in this update:
* Fix trailing return type issue affecting GCC 7.1
* Optimization: acquire BC lists outside the CheckPointIO id loop.
* Add --enable-metaphysicl-required configure option.
* Add configure test for std containers with iterating-returning erase.
* Avoid modifying solution vector in TaoOptimizationSolver callbacks.
* Fix warnings discovered in GCC 8.1 and clang 6.1.
* Fix bug in Parallel::minloc/maxloc.
* Add ability to disable default GhostingFunctors.
* Upgrade libMesh MetaPhysicL submodule to version 0.3.1.
* Add new PetscNonlinearSolver interface: set_snesmf_reuse_base().
* Make SparseMatrix::add() const correct.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 655cf5f0984cd5489592188478eb275901d95a02
+Subproject commit ab2cf97250f90e88b1da3c9fb40931cf46af7ba9
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [App tests PETSc alt](https://civet.inl.gov/job/248254/) on 3b2286f : invalidated by @jwpeterson

Several apps failed the &quot;Pull App&quot; step due to networking issues within HPC. Some of the test diffs may also be real.

moosebuild commented:
Job [Test Intel](https://civet.inl.gov/job/248250/) on 3b2286f : invalidated by @jwpeterson

Intel compiler license server failure while building modules.

jwpeterson commented:
The "Distributed mesh sweep" test failure is only one test that changed recently and the diff is bogus... I will try to push up a fix for that.

jwpeterson committed:
Use absolute tolerances when comparing values close to zero.

This test used to be restricted to one processor but there is no
fundamental reason for this, it can run on up to 4 in my testing.

Refs #12376.

lindsayad commented:
The "App tests PETSc alt" BISON single test failure is exactly the failure we expect. Once the new `Executioner` option is available (once this PR is merged) we will fix that test.

jwpeterson committed:
Set an abs_zero for the test.

This avoids relative error comparisons with numbers that are approximately zero.
